ribbon-eureka/src/main/java/com/netflix/loadbalancer/EurekaDynamicNotificationLoadBalancer.java

```diff
@@ -0,0 +1,159 @@
+                new LegacyEurekaClientProvider(),
+                DEFAULT_SERVER_LIST_UPDATE_EXECUTOR
+        );
+    }
+
+    public EurekaDynamicNotificationLoadBalancer(IClientConfig clientConfig,
+                                                 Provider<EurekaClient> eurekaClientProvider,
+                                                 ExecutorService serverListUpdateService) {
+        super();
+
+        this.eurekaClientProvider = eurekaClientProvider;
+        this.serverListUpdateService = serverListUpdateService;
+        if (clientConfig != null) {
+            initWithNiwsConfig(clientConfig);
+        }
+    }
+
+    @Override
+    public void initWithNiwsConfig(IClientConfig clientConfig) {
+        try {
+            super.initWithNiwsConfig(clientConfig);
+
+            DiscoveryEnabledNIWSServerList myServerList = new DiscoveryEnabledNIWSServerList(
+                    clientConfig,
+                    eurekaClientProvider
+            );
+            setServerListImpl(myServerList);
+
+            AbstractServerListFilter<DiscoveryEnabledServer> myFilter = myServerList.getFilterImpl(clientConfig);
+            myFilter.setLoadBalancerStats(getLoadBalancerStats());
+            setFilter(myFilter);
+
+            connectToEureka();
+
+            LOGGER.info("EurekaDynamicNotificationLoadBalancer for client {} initialized: {}",
+                    clientConfig.getClientName(), this.toString());
+        } catch (Exception e) {
+            throw new RuntimeException("Exception while initializing EurekaDynamicNotificationLoadBalancer," +
+                    " niwsClientConfig: " + clientConfig, e);
+        }
+    }
+
+    /**
+     * Makes no sense to ping an inmemory disc client
+     */
+    @Override
+    public void forceQuickPing() {
+        // no-op
+    }
+
+    @Override
+    public void stopServerListRefreshing() {
+        if (eurekaClientProvider != null) {
+            EurekaClient eurekaClient = eurekaClientProvider.get();
+            if (eurekaClient != null) {
+                eurekaClient.unregisterEventListener(updateListener);
+            }
+        }
+
+        serverListUpdateService.shutdown();
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder("EurekaDynamicNotificationLoadBalancer:");
+        sb.append(super.toString());
+        sb.append("ServerList:" + String.valueOf(serverListImpl));
+        return sb.toString();
+    }
+
+    @Monitor(name="DurationSinceLastUpdateMs", type= DataSourceType.GAUGE)
+    public long getDurationSinceLastUpdateMs() {
+        return System.currentTimeMillis() - lastUpdated.get();
+    }
+
+    // idempotent
+    private void connectToEureka() {
+        updateInitialServerList();
+        if (eurekaClientProvider != null) {
+            EurekaClient eurekaClient = eurekaClientProvider.get();
+            if (eurekaClient != null) {
+                eurekaClient.registerEventListener(updateListener);
```

Comment:
We may want to add a registerEventListener(listener, executor) to the eurekaClient API to take an executor in addition to the listener.